File: AMPPathTracer/Scene.cpp
#include "Common.h"

#include "Scene.h"
#include "LCGRandom.h"

#include <float.h>

V3 CosineWeightedInCone(const V3& normal, LCGRandom *randomGen) restrict(amp, cpu)
{
    float Xi1 = randomGen->NextFloat();
    float Xi2 = randomGen->NextFloat();

    float theta = acosf(sqrtf(Xi1));
    float phi = 2.0f * M_PI * Xi2;

    return normal.RotateByAngles(theta, phi);
}

V3 UniformWeightedInCone(const V3& normal, LCGRandom *randomGen) restrict(amp, cpu)
{
    float Xi1 = randomGen->NextFloat();
    float Xi2 = randomGen->NextFloat();

    float theta = Xi1 * M_PI_OVER_2;
    float phi = 2.0f * M_PI * Xi2;

    return normal.RotateByAngles(theta, phi);
}

V3 Scene::TraceRayCPU(Ray r, LCGRandom *randomGen, int bounces, std::vector<SceneObject> objects) restrict(cpu)
{
    V3 colorAcc = { 0.0f, 0.0f, 0.0f };
    V3 reflectanceAcc = { 1.0f, 1.0f, 1.0f };

    for (int bounce = 0; bounce < bounces; bounce++)
    {
        V3 color;

        // Get nearest intersection
        Intersection nearestI;
        float lengthSqr = FLT_MAX;

        for (unsigned int i = 0; i < objects.size(); i++)
        {
            SceneObject obj = objects[i];
            Intersection tempI = obj.Trace(r);

            if (!tempI.Valid)
                continue;

            float lengthNew = (tempI.Position - r.Origin).LengthSquared();

            if (lengthNew < lengthSqr)
            {
                nearestI = tempI;
                lengthSqr = lengthNew;
            }
        }

        if (nearestI.Valid)
        {
            V3 dir;

            color = { 1.0f, 1.0f, 1.0f };

            V3 normal = nearestI.Normal;

            V3 normal2 = nearestI.Normal.Dot(r.Direction) < 0.0f ? normal : -normal;

            bool inside = normal.Dot(normal2) < 0.0f;

            float reflectivity = nearestI.Object.Properties.Reflectivity;

            float rValue = randomGen->NextFloat();

            if (rValue < reflectivity)
            {
                dir = r.Direction.Reflect(normal);

                // Specular color to approximate metals that tint reflections
                color = nearestI.Object.Properties.SpecularColor;
            }
            else
            {
                dir = CosineWeightedInCone(nearestI.Normal, randomGen);

                // Diffuse reflection affected by surface color
                color = nearestI.Object.Properties.DiffuseColor;
            }

            colorAcc += reflectanceAcc * nearestI.Object.Properties.Emission;
            reflectanceAcc *= color;

            r.Origin = nearestI.Position + dir * 1E-3f;
            r.Direction = dir;
        }
        else
        {
            break;
        }
    }

    return colorAcc;
}

V3 Scene::TraceRay(Ray r, LCGRandom *randomGen, int bounces, array<SceneObject> &objects) restrict(amp)
{
    V3 colorAcc = { 0.0f, 0.0f, 0.0f };
    V3 reflectanceAcc = { 1.0f, 1.0f, 1.0f };

    for (int bounce = 0; bounce < bounces; bounce++)
    {
        V3 color;

        // Get nearest intersection
        Intersection nearestI;
        float lengthSqr = FLT_MAX;

        for (unsigned int i = 0; i < objects.extent.size(); i++)
        {
            SceneObject obj = objects[index<1>(i)];
            Intersection tempI = obj.Trace(r);

            if (!tempI.Valid)
                continue;

            float lengthNew = (tempI.Position - r.Origin).LengthSquared();

            if (lengthNew < lengthSqr)
            {
                nearestI = tempI;
                lengthSqr = lengthNew;
            }
        }

        if (nearestI.Valid)
        {
            V3 dir;

            color = { 1.0f, 1.0f, 1.0f };

            V3 normal = nearestI.Normal;

            V3 normal2 = nearestI.Normal.Dot(r.Direction) < 0.0f ? normal : -normal;

            bool inside = normal.Dot(normal2) < 0.0f;

            float reflectivity = nearestI.Object.Properties.Reflectivity;

            float rValue = randomGen->NextFloat();

            if (rValue < reflectivity)
            {
                dir = r.Direction.Reflect(normal);

                // Specular color to approximate metals that tint reflections
                color = nearestI.Object.Properties.SpecularColor;
            }
            else
            {
                dir = CosineWeightedInCone(nearestI.Normal, randomGen);

                // Diffuse reflection affected by surface color
                color = nearestI.Object.Properties.DiffuseColor;
            }

            colorAcc += reflectanceAcc * nearestI.Object.Properties.Emission;
            reflectanceAcc *= color;

            r.Origin = nearestI.Position + dir * 1E-3f;
            r.Direction = dir;
        }
        else
        {
            break;
        }
    }

    return colorAcc;
}

File: AMPPathTracer/scene.lua
Scene = {
	Threshold = 0,
	Recursions = 10,
	SceneEmission = { 0, 0, 0 },
	Camera = {
		Fov = 1.3962634016,
		Yaw = 0,
		FocalPoint = 10,
		Aperture = 0.01,
		Origin = { 0, 0, 0 },
		Pitch = 0,
	},
	Objects = {
		{
			Type = 1,
			Point = { 0, 0, -25 },
			Normal = { 0, 0, 1 },
			Properties = {
				SpecularColor = { 1, 1, 1 },
				DiffuseColor = { 0.75, 0.75, 0.75 },
				RefractiveIndex = 1,
				Emission = { 0, 0, 0 },
				Transmittance = 0,
				Reflectivity = 0,
				Roughness = 0,
			},
		},
		{
			Type = 1,
			Point = { 0, 0, 25 },
			Normal = { 0, 0, -1 },
			Properties = {
				SpecularColor = { 1, 1, 1 },
				DiffuseColor = { 0.75, 0.75, 0.75 },
				RefractiveIndex = 1,
				Emission = { 0, 0, 0 },
				Transmittance = 0,
				Reflectivity = 0,
				Roughness = 0,
			},
		},
		{
			Type = 1,
			Point = { 0, 100, 0 },
			Normal = { 0, -1, 0 },
			Properties = {
				SpecularColor = { 1, 1, 1 },
				DiffuseColor = { 0.75, 0.75, 0.75 },
				RefractiveIndex = 1,
				Emission = { 0, 0, 0 },
				Transmittance = 0,
				Reflectivity = 0,
				Roughness = 0,
			},
		},
		{
			Type = 1,
			Point = { 0, -5, 0 },
			Normal = { 0, 1, 0 },
			Properties = {
				SpecularColor = { 1, 1, 1 },
				DiffuseColor = { 0.75, 0.75, 0.75 },
				RefractiveIndex = 1,
				Emission = { 0, 0, 0 },
				Transmittance = 0,
				Reflectivity = 0,
				Roughness = 0,
			},
		},
		{
			Type = 1,
			Point = { 25, 0, 0 },
			Normal = { -1, 0, 0 },
			Properties = {
				SpecularColor = { 1, 1, 1 },
				DiffuseColor = { 0.25, 0.75, 0.25 },
				RefractiveIndex = 1,
				Emission = { 0, 0, 0 },
				Transmittance = 0,
				Reflectivity = 0,
				Roughness = 0,
			},
		},
		{
			Type = 1,
			Point = { -25, 0, 0 },
			Normal = { 1, 0, 0 },
			Properties = {
				SpecularColor = { 1, 1, 1 },
				DiffuseColor = { 0.75, 0.25, 0.25 },
				RefractiveIndex = 1,
				Emission = { 0, 0, 0 },
				Transmittance = 0,
				Reflectivity = 0,
				Roughness = 0,
			},
		},
		{
			Type = 0,
			Radius = 500,
			Origin = { 0, 60, 524.9 },
			Properties = {
				SpecularColor = { 1, 1, 1 },
				DiffuseColor = { 0.75, 0.75, 0.75 },
				RefractiveIndex = 1,
				Emission = { 4, 4, 4 },
				Transmittance = 0,
				Reflectivity = 0,
				Roughness = 0,
			},
		},
		{
			Type = 0,
			Radius = 10,
			Origin = { 12, 65, -15 },
			Properties = {
				SpecularColor = { 1, 1, 1 },
				DiffuseColor = { 0.75, 0.75, 0.75 },
				RefractiveIndex = 1.4,
				Emission = { 0, 0, 0 },
				Transmittance = 1,
				Reflectivity = 1,
				Roughness = 0,
			},
		},
		{
			Type = 0,
			Radius = 10,
			Origin = { -12, 80, -15 },
			Properties = {
				SpecularColor = { 0.9, 0.75, 0 },
				DiffuseColor = { 0.75, 0.75, 0.75 },
				RefractiveIndex = 50,
				Emission = { 0, 0, 0 },
				Transmittance = 0,
				Reflectivity = 1,
				Roughness = 0.1,
			},
		},
	},
}

return Scene